fooof/Compute_broadbandpower_offset_slope.py

```diff
@@ -0,0 +1,347 @@
+subjects = pd.read_csv(subject_list, delimiter=';', header=0, usecols=[0,1])
+subjects['Case_ID'] = subjects['Case_ID'].apply(lambda x: str(x).zfill(4)) # add padding zeros when necessary
+
+# get location of ASCIIs of MEG 
+subjects['location_MEG'] = dir_input + 'sub-'+ subjects['Case_ID'] + '/meg/' + subjects['MM'] + '/' + atlas + '/' 
+
+
+###### Compute power spectrum ######
+## takes a while when multiple subjects are analyzed at once
+mean_pxx, broadband_power, f = run_loop_powerspectrum(subjects, nr_epochs=nr_epochs, nr_rois=nr_rois, 
+                                                      Fs=Fs, window_length=4096, freq_range=freq_range)
+       
+# mean_pxx contains the average power spectra over nr_epochs for all subjects for all rois
+# broadband_power is the area under the average power spectra over the frequency range for all subjects and rois
+# f gives the frequencies of the power spectrum (can be useful when plotting power spectrum)
+
+# save output
+if save_output == True:
+    np.save(dir_output + name_to_save + '_mean_pxx', mean_pxx)
+    np.save(dir_output + name_to_save + '_broadband_power', broadband_power)
+    print('saving power spectra and broadband power')
+    
+
+
+
+###### Compute FOOOOF offset and slope ######
+# create empty arrays to store offset and slope values
+offset = np.empty([mean_pxx.shape[0], mean_pxx.shape[2]]) # index 0 -> number of subjects, index 2 -> number of rois
+slope = np.empty([mean_pxx.shape[0], mean_pxx.shape[2]])
+
+print('Running FOOOF.........................')
+
+# run across all subjects in your list
+for sub in range(len(subjects)):
+    print('subject : ' + str(sub)) # print which subject is analyzed
+    if np.isnan(mean_pxx[sub,1,0]): # if there is no mean_pxx present for a subject, set offset/slope to nan
+        offset[sub,] = np.nan
+        slope[sub,] = np.nan
+    else:
+        i=0
+        for roi in nr_rois:
+            print('roi : ' + str(roi))
+            offset[sub,i], slope[sub,i] = cal_FOOOF_parameters(mean_pxx[sub,:,i], f, freq_range=[0.5, 48])
+            i=i+1
+            time.sleep(0.05) # by having a pause in the analyses it should limit the %CPU continuously
+    # might be relevant when computing FOOOF for large dataset        
+    if sub % 10 ==0 & save_output==True:
+        print('save offset and slope in between after (another) 10 subjects')
+        np.save(dir_output + name_to_save + '_FOOOF_offset', offset)
+        np.save(dir_output + name_to_save + '_FOOOF_slope', slope)
+    
+if save_output == True:          
+    np.save(dir_output + name_to_save + '_FOOOF_offset', offset)
+    np.save(dir_output + name_to_save + '_FOOOF_slope', slope)
+    # save offset and slope output of FOOOF
+    print("FOOOF-offset is saved: "+ dir_output + name_to_save + "_FOOOF_offset" )
+    print("FOOOF-slope is saved: "+ dir_output + name_to_save + "_FOOOF_slope" )
+    
+######## Combine results with Case_ID in 'subjects' variable
```

Comment:
is this the same "heading level" as ###### Compute FOOOOF offset and slope ###### ? If it is, would suggest putting the ### at the end as well for consistency.

Reply:
yes, added it